app/controllers/tasks_controller.rb

```diff
@@ -0,0 +1,71 @@
+require 'date'
+
+class TasksController < ApplicationController
+
+  def index
+    @tasks = Task.all
+  end
+
+  def show
+    task_id = params[:id]
+    @task = Task.find_by(id: task_id)
+
+    if @task.nil?
+      head 404
+    end
+  end
+
+  def new
+    @task = Task.new
+  end
+
+  def create
+    @task = Task.new(
+      task_params
+    )
+
+    if @task.save
+      redirect_to tasks_path
+    else
+      render :new
+    end
+  end
+
+  def edit
+    @task = Task.find(params[:id])
+    if @task.date_completed
+      redirect_to :action => 'index', :alert => "Cannot update a completed task!"
+    end
+  end
+
+  def completed
+    today_date = Date.today
+    @task = Task.find(params[:id])
+
+    @task.update(completed: true, date_completed: today_date)
+    redirect_to root_path
+  end
+
+  def update
+    @task = Task.find(params[:id])
+
+    @task.update(action: params[:task][:action], description: params[:task][:description], completion_date: params[:task][:completion_date])
```

Comment:
This is pretty correct! However, a shorter and more secure way to do this would be to use the `task_params` method you used below. Much like how you used task_params in `create` on line 23, you could do here:
```ruby
@task.update(task_params)
```